Add new page for Cross-compilation (#29)
ros2/ros2_documentation: Cross-compilation.rst

 - ``CMAKE_FIND_ROOT_PATH_MODE_*``: the search strategy for program,package,library, and include, usually: ``NEVER`` (look on the host-fs), ``ONLY`` (look on sysroot), ``ONLY`` (look on sysroot) and ``ONLY`` (look on sysroot)
 - ``PYTHON_SOABI``: the index name of the python libraries generated by ROS2, e.g. ``cpython-36m-aarch64-linux-gnu``
 - ``THREADS_PTHREAD_ARG "0" CACHE STRING "Result from TRY_RUN" FORCE``: Force the result of the ``TRY_RUN`` cmd to 0 (success) because binaries can not run on the host system.

The toolchain-file is provided to CMake with the ``-DCMAKE_TOOLCHAIN_FILE=path/to/file`` parameter. This will also set the ``CMAKE_CROSSCOMPILING`` variable to ``true`` which can be used by the software being built.

The ``CMAKE_SYSROOT`` is particularly important for ROS2 as the packages need many dependencies (e.g. python, openssl, opencv, poco, eigen3, ...).
Setting ``CMAKE_SYSROOT`` to a target file-system with all the dependencies installed on it will allow CMake to find them during the cross-compilation.

.. note:: You can find more information on the CMake `documentation <https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html>`__ page.

> When dowloading the ROS2 source code, a generic toolchain-file is available in ``src/ros2/cross_compile/cmake-toolchains/generic_linux.cmake``. Further examples on using it can be found on the `Cross-compiling examples for Arm`_ section.

Reviewer: ```suggestion
When dowloading the ROS2 source code, a generic toolchain-file is available in the repository https://github.com/ros2/cross_compile which can be downloaded separately. Further examples on using it can be found on the `Cross-compiling examples for Arm`_ section.
```

Target file-system
------------------

As mentioned previously, ROS2 requires different libraries which needs to be provided to cross-compile.

There are a number of ways to obtain the file-system:
 - downloading a pre-built image
 - installing the dependencies on the target and exporting the file-system (e.g. with sshfs)
 - using qemu + docker (or chroot) to generate the file-system on the host machine.

.. note:: You can find information on how to use Docker + qemu on the next `Cross-compiling examples for Arm`_ section.

Build process
-------------

> The build process is similar to native compilation. The only difference is an extra argument to ``Colcon`` to specify the ``toochainfile``:

Reviewer: ```suggestion
The build process is similar to native compilation. The only difference is an extra argument to ``Colcon`` to specify the ``toolchainfile``:
```

spelling

.. code-block:: bash

    colcon build --merge-install \
        --cmake-force-configure \
        --cmake-args \
            -DCMAKE_TOOLCHAIN_FILE="<path_to_toolchain/toolchainfile.cmake>"

The ``toolchain-file`` provide to CMake the information of the ``cross-compiler`` and the ``target file-system``.
``Colcon`` will call CMake with the given toolchain-file on every package of ROS2.

Cross-compiling examples for Arm
================================

> When `downloading the ROS2 source code <https://index.ros.org/doc/ros2/Linux-Development-Setup/>`__, a number of cross-compiling assets are available under the src/ros2/cross_compile directory.

Reviewer: In https://github.com/ros2/ros2/pull/616 I've suggested not including the cross_compile repository in the default repos file. Which means that these instructions need an update for cloning it separately.

```suggestion
After `downloading the ROS2 source code <https://index.ros.org/doc/ros2/Linux-Development-Setup/>`__, you can add cross-compilation assets to the workspace via git: ``git clone https://github.com/ros2/cross_compile.git src/ros2/cross_compile`.
```

These are working examples on how to cross-compile for Arm cores.

The following targets are supported:
 - Ubuntu-arm64: To be used with any ARMv8-A based system.
 - Ubuntu-armhf: To be used with any modern ARMv7-A based system.

> There are 4 main steps:

Reviewer: ```suggestion
Therse are the main steps:
```

 - Installing development tools

>  - Downloading ROS2 source code

Reviewer: ```suggestion
 - Downloading ROS2 source code
 - Downloading the ROS2 cross compilation assets
```

The package ``class-loader`` has a known issue where it is searching for ``libz`` and ``libpcre`` on the host system instead of SYSROOT. As a workaround for the moment, please link both libraries into the the host's file-system

.. code-block:: bash

    mkdir -p /usr/lib/$TARGET_TRIPLE
    ln -s `pwd`/sysroot_docker/lib/$TARGET_TRIPLE/libz.so.1 /usr/lib/$TARGET_TRIPLE/libz.so
    ln -s `pwd`/sysroot_docker/lib/$TARGET_TRIPLE/libpcre.so.3 /usr/lib/$TARGET_TRIPLE/libpcre.so

Then, start a build with colcon specifying the toolchain-file:

.. code-block:: bash

    cd ros2_ws

    colcon build --merge-install \
        --cmake-force-configure \
        --cmake-args \
            -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON \
            -DCMAKE_TOOLCHAIN_FILE="$(pwd)/src/ros2/cross_compile/cmake-toolchains/generic_linux.cmake" \
            -DSECURITY=ON

Done! The install and build directories will contain the cross-compiled assets.

Automated Cross-compilation
===========================

All the steps above are also included into a Dockerfile and can be used for automation/CI.

First, download the dockerfile and build the image:

.. code-block:: bash

>     wget https://raw.githubusercontent.com/ros2-for-arm/cross_compile/Dockerfile_cc_for_arm

Reviewer: ```suggestion
    wget https://raw.githubusercontent.com/ros2/cross_compile/Dockerfile_cc_for_arm
```

Update to the new repository location
